File: python/word-wrap/WordWrapperTests.py
import unittest

class WordWrapperTest(unittest.TestCase):
	def test_should_return_same_exact_text(self):
		self.assertEqual("Test", WordWrapper.wrapText("Test", 4))
	
	def test_should_break_word(self):
		self.assertEqual("Te\nst", WordWrapper.wrapText("Test", 2))
	
	def test_should_break_word_twice(self):
		self.assertEqual("Tes\ntin\ng", WordWrapper.wrapText("Testing", 3))
	
	def test_should_replace_space(self):
		self.assertEqual("Test\ntest", WordWrapper.wrapText("Test test", 4))
	
	def test_should_replace_previous_space(self):
		self.assertEqual("Test\ntest", WordWrapper.wrapText("Test test", 7))

class WordWrapper:
	@staticmethod
	def wrapText(text, maxLength):
		wrapper = WordWrapper(maxLength)
		return wrapper.wrap(text)
	
	def __init__(self, maxLength):
		self.maxLength = maxLength
	
	def wrap(self, text):
		if(len(text) <= self.maxLength):
			return text
		if(text[self.maxLength] == " "):
			return self.break_line(text, self.maxLength, 1)
		elif(text.find(" ") > 0):
			return self.break_line(text, text.find(" "), 1)
		else:
			return self.break_line(text, self.maxLength, 0)
	
	def break_line(self, text, index, gap):
		return text[0 : index] + "\n" + self.wrap(text[index + gap : len(text)])

if __name__ == '__main__':
	unittest.main()

File: ruby/word-wrap/WordWrapperTests.rb
require "test/unit"

class WordWrapperTests < Test::Unit::TestCase
  def test_should_return_exact_same_word
    assert_equal("word", WordWrapper::wrap("word", 4))
  end
  
  def test_should_break_word_once
    assert_equal("wo\nrd", WordWrapper::wrap("word", 2))
  end
  
  def test_should_break_word_twice
    assert_equal("tes\ntin\ng", WordWrapper::wrap("testing", 3))
  end
  
  def test_should_break_replacing_space
    assert_equal("word\ntest", WordWrapper::wrap("word test", 4))
  end
  
  def test_should_break_replacing_previous_space
    assert_equal("word\ntest", WordWrapper::wrap("word test", 7))
  end
end

class WordWrapper
  def WordWrapper::wrap(text, max_length)
    wrapper = WordWrapper.new(max_length)
    return wrapper.wrap(text)
  end
  
  def initialize(max_length)
    @max_length = max_length
  end
  
  def wrap(text)
    if(text.length() <= @max_length)
      return text
    end
    if(text[@max_length] == " ")
      return break_line_deleting_last_char_at(text, @max_length)
    elsif(text.index(" "))
      return break_line_deleting_last_char_at(text, text.index(" "))
    else
      return break_line_at(text, @max_length)
    end
  end
  
  def break_line_deleting_last_char_at(text, index)
    return break_line(text, index, 1)
  end
  
  def break_line_at(text, index)
    return break_line(text, index, 0)
  end
  
  def break_line(text, index, gap)
    return text[0..index-1] + "\n" + wrap(text[index+gap..text.length()])
  end
end